cpp/box/Box.h

```diff
@@ -343,6 +347,53 @@ class Box
         });
     }
 
+    //! Compute center of mass for vectors
+    /*! \param vecs Vectors to compute center of mass
+     *  \param Nvecs Number of vectors
+     *  \param masses Optional array of masses, of length Nvecs
+     *  \return Center of mass as a vec3<float>
+     */
+    vec3<float> centerOfMass(vec3<float>* vecs, unsigned int Nvecs, float* masses = NULL) const
+    {
+        // This roughly follows the implementation in
+        // https://en.wikipedia.org/wiki/Center_of_mass#Systems_with_periodic_boundary_conditions
+        float total_mass(0);
+        vec3<std::complex<float>> xi_mean(vec3<std::complex<float>>(0.0, 0.0, 0.0));
+
+        for (size_t i = 0; i < Nvecs; ++i) {
+            vec3<float> phase(TWO_PI * makeFractional(vecs[i]));
+            vec3<std::complex<float>> xi(
+                    std::polar(float(1.0), phase.x),
+                    std::polar(float(1.0), phase.y),
+                    std::polar(float(1.0), phase.z));
+            float mass = (masses != NULL) ? masses[i] : 1.0;
+            total_mass += mass;
+            xi_mean += std::complex<float>(mass, 0) * xi;
+        }
+        xi_mean /= std::complex<float>(total_mass, 0);
+
+        return wrap(makeAbsolute(vec3<float>(
+                std::arg(xi_mean.x),
+                std::arg(xi_mean.y),
+                std::arg(xi_mean.z)) / TWO_PI));
+    }
```

Comment:
According to [the documentation](https://en.cppreference.com/w/cpp/numeric/math/atan2) for `atan2` (which governs the error cases for `std::arg`), whether or not `std::arg(std::complex(0))` throws a domain error is implementation-specific. I tested on my Mac and the vislab and it seems fine, but it might be worth either checking if any of the components of `xi_mean` are 0 to avoid NaNs here (or use the actual error-checking functionality as documented [here](https://en.cppreference.com/w/cpp/numeric/math/math_errhandling).

Reply:
@vyasr I don't understand what you want me to do here. I read this [page about atan2](https://en.cppreference.com/w/cpp/numeric/math/atan2) but I'm not sure how to proceed. I would appreciate help.

Reply:
The page you linked seems to indicate that `std::atan2(0, 0) = 0` and is not implementation-dependent...?

Reply:
I'm reading the lines under "Error handling" on that page. The second line states that "Domain error may occur if x and y are both 0". To see how to handle this properly, look at the documentation of `math_errhandling` [here](https://en.cppreference.com/w/cpp/numeric/math/math_errhandling). I totally appreciate that it looks gross, but this exact kind of problem was causing hard-to-find bugs for us in both Steinhardt and LocalDescriptors, so I'd like to nip that in the bud here rather than risk it cropping up in an unexpected manner later.

Reply:
Feel free to find me tomorrow or just Slack me if you want more specific guidance, I'm happy to explain in more detail.

Reply:
(summary of Slack discussion) This is fine since we're assuming IEEE 754 compliance.

Reply:
After independent discussion, we concluded that:
1. Supporting non-IEEE 754 compliant systems is out of scope and unnecesary.
2. In the specific edge-cases that would be problematic (i.e. perfectly distributed systems where the values of `xi_mean` are 0), there is no "correct" answer for the center of mass, so rather than trying to impose our own convention we'll just accept whatever the standard library returns. This means that based on the exact floating point number provided, we could get cases like `std::arg(std::complex<float>(+0, -0))`, and that will be resolved according to the definition of `std::atan2`.